### PR TITLE
Allow non valid zones for instances to be add to a placement

### PR DESCRIPTION
**Pull request by nerd0**
There are certain scenarios were instances within a placement will not be apart of the same zone of the placement. This PR relaxes the constraint a bit by allowing a `placement.Option` to bypass the strict check.

### CODE REVIEW

File: src/cluster/placement/options.go [hunk 1]
```diff
@@ -68,6 +71,7 @@ type options struct {
 	validateFn          ValidateFn
 	nowFn               clock.NowFn
 	allowPartialReplace bool
+	allowNonValidZones  bool
```
Comment (nerd0) [inline src/cluster/placement/options.go]: Not a fan of the name, perhaps `allowAnyZone` ?

Comment (arnikola) [reply src/cluster/placement/options.go]: 'allowAllZones` maybe?

File: src/cluster/placement/selector/common.go [hunk 2]
```diff
@@ -67,16 +71,18 @@ func filterZones(
 	}
 
 	var validZone string
+	var allowNonValidZone bool
 	if opts != nil {
 		validZone = opts.ValidZone()
+		allowNonValidZone = opts.AllowNonValidZones()
```
Comment (arnikola) [inline src/cluster/placement/selector/common.go]: I think you can just do 
```golang
// ...
if opts.AllowNonValidZones() {
  return candidates
}
```

File: src/cluster/placement/options.go (hunk 1 above)
Comment (nerd0) [reply src/cluster/placement/options.go]: Good call.

File: src/cluster/placement/selector/common.go [hunk 3]
```diff
@@ -67,16 +71,18 @@ func filterZones(
...
 	}
 	if validZone == "" && len(p.Instances()) > 0 {
 		validZone = p.Instances()[0].Zone()
 	}
 
 	validInstances := make([]placement.Instance, 0, len(candidates))
 	for _, instance := range candidates {
-		if validZone == instance.Zone() {
+		if validZone == instance.Zone() || allowNonValidZone {
```
[the remark this answers is not in the document]
Comment (nerd0) [reply src/cluster/placement/selector/common.go]: Good point but I think your suggestion below is a bit more streamlined.

File: src/cluster/placement/selector/common.go (hunk 2 above)
Comment (nerd0) [reply src/cluster/placement/selector/common.go]: Great idea, thanks for the feedback

File: src/cluster/placement/selector/common.go [hunk 4]
```diff
@@ -46,6 +46,10 @@ func getValidCandidates(
 		}
 		if instanceInPlacement.IsLeaving() {
 			instances = append(instances, instanceInPlacement)
+			continue
+		}
+		if opts.ShardStateMode() == placement.StableShardStateOnly {
```
Comment (jeromefroe) [inline src/cluster/placement/selector/common.go]: Talked offline, going to remove this check for now until we have a more concrete use case for it.

**Pull request merged**